File: cache/src/main/java/com/mm/http/Converter.java
package com.mm.http;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

import javax.annotation.Nullable;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.http.Body;
import retrofit2.http.Part;
import retrofit2.http.PartMap;

/**
 * Created by : majian
 * Date : 2021/8/24
 */

public interface Converter<F, T> {
    @Nullable
    T convert(F value) throws IOException;

    /** Creates {@link Converter} instances based on a type and target usage. */
    abstract class Factory {

        public @Nullable
        Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations, RetrofitCache retrofit) {
            return null;
        }

        /**
         * Returns a {@link Converter} for converting {@code type} to an HTTP request body, or null if
         * {@code type} cannot be handled by this factory. This is used to create converters for types
         * specified by {@link Body @Body}, {@link Part @Part}, and {@link PartMap @PartMap} values.
         */
        public @Nullable
        Converter<?, RequestBody> requestBodyConverter(
                Type type,
                Annotation[] parameterAnnotations,
                Annotation[] methodAnnotations,
                RetrofitCache retrofit) {
            return null;
        }

        /**
         * Extract the upper bound of the generic parameter at {@code index} from {@code type}. For
         * example, index 1 of {@code Map<String, ? extends Runnable>} returns {@code Runnable}.
         */
        protected static Type getParameterUpperBound(int index, ParameterizedType type) {
            return Utils.getParameterUpperBound(index, type);
        }

        /**
         * Extract the raw class type from {@code type}. For example, the type representing {@code
         * List<? extends Runnable>} returns {@code List.class}.
         */
        protected static Class<?> getRawType(Type type) {
            return Utils.getRawType(type);
        }
    }
}

File: retrofit_cache/src/main/java/com/mm/http/Converter.kt
package com.mm.http

import okhttp3.RequestBody
import okhttp3.ResponseBody
import java.io.IOException
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

/**
 * 缓存 Converter
 */
interface Converter<F, T> {

    @Throws(IOException::class)
    fun convert(value: F): T

    /** Creates [Converter] instances based on a type and target usage.  */
    abstract class Factory {

        open fun <T> responseBodyConverter(
            type: Type,
            annotations: Array<Annotation>?,
            retrofit: RetrofitCache
        ): Converter<ResponseBody, T>? {
            return null
        }

        /**
         * Returns a [Converter] for converting `type` to an HTTP request body, or null if
         * `type` cannot be handled by this factory. This is used to create converters for types
         * specified by [@Body][Body], [@Part][Part], and [@PartMap][PartMap] values.
         */
        open fun requestBodyConverter(
            type: Type, parameterAnnotations: Array<Annotation>?,
            methodAnnotations: Array<Annotation>?, retrofit: RetrofitCache
        ): Converter<*, RequestBody>? {
            return null
        }

        companion object {
            /**
             * Extract the upper bound of the generic parameter at `index` from `type`. For
             * example, index 1 of `Map<String, ? extends Runnable>` returns `Runnable`.
             */
            protected fun getParameterUpperBound(index: Int, type: ParameterizedType): Type {
                return type.getParameterUpperBound(index)
            }

            /**
             * Extract the raw class type from `type`. For example, the type representing `List<? extends Runnable>` returns `List.class`.
             */
            protected fun getRawType(type: Type): Class<*> {
                return type.getRawType()
            }
        }
    }
}